File: backend/app/schemas/category.py
"""
    Schemas for /categories/ API.
"""
from typing import Optional
from pydantic import BaseModel, field_validator
from models.category import *
from schemas.article import ArticleOutput, ArticlePreview
from schemas.category_preview import CategoryPreview
import re

INVALID_URL_PATTERN = re.compile(r"[^\w]", re.A) # Catches non-alphanumeric characters (including non-ASCII), except underscore.

class CategoryInput(BaseModel):
    """Schema for creation requests."""
    name: str # User-friendly title
    description: Optional[str] = ""
    directory_name: str # The last URL component of the path
    parent_category_path: str # Path to parent category
    
    @field_validator("directory_name", check_fields=False)
    def validate_url(cls, url: str):
        if INVALID_URL_PATTERN.search(url): # Cannot contain slashes or characters that are reserved or would require url-encoding
            raise ValueError("Invalid directory name")

        return url
    
    @field_validator("parent_category_path", check_fields=False)
    def validate_category_parent_path(cls, path: str):
        # Prevent creating additional root categories
        if path.find("//") != -1 or (len(path) > 1 and path[-1] == "/") or (path == "" or path[0] != "/"): # Cannot have 2 consecutive slashes or end in a slash (except the root path "/")
            raise ValueError("Invalid path")
        return path
    
class CategoryUpdate(BaseModel):
    """Schema for patching requests."""
    name: Optional[str] = None
    description: Optional[str] = None
    directory_name: Optional[str] = None
    view_type: Optional[CategoryViewEnum] = None
    sorting_type: Optional[CategorySortingModeEnum] = None
    parent_category_path: Optional[str] = None

class CategoryOutput(CategoryPreview):
    """Schema with a complete view of the category's data."""
    articles: list[ArticlePreview]
    total_articles: int # Total amount of articles regardless of pagination limits applied to the articles field.
    subcategories: list["CategoryOutput"] # Child categories, recursive.

File: frontend/services/category.ts
/**
 * Utility methods for /categories/ API routes.
 */
import type { ArticlePreview } from './article'
import Service from './service'

export type categoryID = integer
export type categorySortingMode = "chronological"|"manual"
export type categoryViewMode = "vertical"|"grid"

export type CategoryPreview = {
  id: categoryID,
  name: string,
  description: string,
  directory_name: string,
  path: path,
  view_type: categoryViewMode,
  sorting_type: categorySortingMode,
}

export type Category = CategoryPreview & {
  path: string,
  articles: ArticlePreview[],
  /** Total amount of articles regardless of pagination limits applied to the articles field. */
  total_articles: integer,
  subcategories: Category[],
}

export type CategoryCreationRequest = {
  name: string,
  description: string,
  directory_name: string,
  parent_category_path: string,
}


export type CategoryUpdateRequest = {
  name?: string,
  description?: string,
  directory_name?: string,
  view_type?: categoryViewMode,
  sorting_type?: categorySortingMode,
  parent_category_path?: string,
}

class CategoryService extends Service {
  constructor(api_url: string) {
    super(api_url)
  }

  /** Creates a category. */
  async createCategory(request: CategoryCreationRequest): Promise<Category> {
    const response = await this.post('/categories/', request)
    return response.data
  }

  /** Fetches a category by its path; expects a leading slash. */
  async getCategory(path: string, published_only?: boolean, limit?: integer, skip?: integer): Promise<Category> {
    const response = await this.get('/categories' + path, {
      articles_amount: limit,
      articles_skip: skip,
      published_only: published_only,
    })
    return response.data
  }

  /** Updates a category's data. */
  async updateCategory(path: string, request: CategoryUpdateRequest): Promise<Category> {
    const response = await this.patch('/categories' + path, request) // Path is expected to have leading slash.
    return response.data
  }
}

export default CategoryService